src/core/connection.c

```diff
@@ -109,6 +109,7 @@ QuicConnAlloc(
         Connection->Stats.CorrelationId);
 
     Connection->RefCount = 1;
+    Connection->ExternalRefCount = IsServer ? 1 : 2;
```

Comment:
I disagree with this model. An app that creates a new object should always get an object with a single ref. For instance, if I call `ConnectionOpen` followed by `Release` on it, that should be enough. If (and only if) I need to have additional threads, beyond the QUIC worker, then I call AddRef for each additional dependency.

Reply:
Then this is pointless. You HAVE to do this, otherwise you still have the race conditions described.

Reply:
The whole problem is the existing model basically being a ref count of 1. The APP then has to do the work to synchronize between the worker thread and itself. Thats no different then what we have currently, as the app would still have to manually handle that synchronization.

Reply:
It all depends on the application scenario/usage. If an app calls `ConnectionOpen` followed simply by `ConnectionStart` and then everything happens on the callback thread, they should be able to simply call `Release` once in the callback and be done.

If they want to continue to use the connection on the main thread after calling start, then they need to call `AddRef` before calling `ConnectionStart`.

I don't understand why that doesn't work, or is an unreasonable design.

Reply:
Because we've seen before you can get a ConnectionShutdown event before ConnectionOpen even returns to the app. And then if you call Release in that callback, you're getting a handle back to your app that is already freed.

Reply:
If you could guarantee no callbacks would ever be called until ConnectionStart has been called, yes you could start with a ref count of 1. But because you can't guarantee that, you have to start with a ref count of 2. Its just object ownership semantics. The same reason why every caller that currently calls us makes sure there is a ref count of 2 on their external ref count before calling ConnectionOpen. A ref count of 1 is fundamentally incorrect here. It doesn't matter that events are unlikely to happen before ConnectionStart. The fact that they can forces the semantics here. Its just a fact of how we designed the API and callback structure.

Reply:
We should definitely talk about this some more. Perhaps we can/should prevent the callbacks before `ConnectionStart`, as I suspect most existing code will break/crash if the connection shutdown complete before that. But that being said, that can only happen under certain (kind of invalid) app behavior of trying to shutdown the registration at the same time as creating a new connection. Let's talk some more.

Reply:
Ideally ConnectionOpen/StreamOpen can *never* fail asynchronously; This starts to look more like ConnectionCreate/StreamCreate, since it will only create the object, and not do anything network side. The alternative is that anything that can invalidate asynchronously never gets returned from the caller, only given in a callback.

Reply:
#2226 is solving the async case. Also open never hits the network anyway. It's basically equivalent to alloc. The only way for it to async fail was another thread shutting down the registration. Which we're solving in that other pr before this.